src/main/scala/com/codacy/parsers/XmlReportParser.scala

```diff
@@ -0,0 +1,55 @@
+package com.codacy.parsers
+
+import java.io.File
+
+import com.codacy.api.CoverageReport
+import com.codacy.parsers.util.XMLoader
+
+import scala.util.{Failure, Success, Try}
+import scala.xml.{Elem, NodeSeq}
+
+trait XmlReportParser {
+  def validateSchema(xml: Elem): Boolean
+
+  def getRootNode(xml: Elem): NodeSeq
+
+  def loadXmlReport(reportFile: File, schemaErrorMessage: String): Either[String, NodeSeq] = {
+    Try(XMLoader.loadFile(reportFile)) match {
+      case Success(xml) if validateSchema(xml) =>
+        Right(getRootNode(xml))
+
+      case Success(_) =>
+        Left(s"Invalid report. $schemaErrorMessage.")
+
+      case Failure(ex) =>
+        Left(s"Unparseable report. ${ex.getMessage}")
+    }
+  }
+
+  private def getFailedParseMessage(ex: Throwable) =
+    s"Failed to parse report with error: ${ex.getMessage}"
+
+  def parseReport(reportFile: File, schemaErrorMessage: String)(
+      parseReport: NodeSeq => CoverageReport
+  ): Either[String, CoverageReport] = {
+    loadXmlReport(reportFile, schemaErrorMessage)
+      .flatMap { report =>
+        Try(parseReport(report)) match {
+          case Success(coverageReport) => Right(coverageReport)
+          case Failure(ex) => Left(getFailedParseMessage(ex))
+        }
+      }
+  }
+
+  def parseReportWithEither(reportFile: File, schemaErrorMessage: String)(
+      parseReport: NodeSeq => Either[String, CoverageReport]
+  ): Either[String, CoverageReport] = {
+    loadXmlReport(reportFile, schemaErrorMessage)
+      .flatMap { report =>
+        Try(parseReport(report)) match {
+          case Success(coverageReport) => coverageReport
+          case Failure(ex) => Left(getFailedParseMessage(ex))
+        }
+      }
+  }
```

Comment:
We should only have one way (`parseReport: NodeSeq => Either[String, CoverageReport]`).

Reply:
I removed one of them and made changed the parseReports that were returning a CoverageReport to return a Right(CoverageReport).
Maybe not the best approach. But let's build on top of this.
@lolgab thoughts?

Reply:
If we don't manage all possible failing cases, I'd say to wrap in a `Try` or `try {} catch {}` before returning. An Either returning function should never fail with exceptions by contract.

Reply:
That's what we are doing. Calling the parseReport wrapped in a `Try()`.
However, to avoid having two `parseReport`s, some parsers are always returning `Right(...)`. Which is weird, but maybe not a problem